### ISSUE TITLE
How to measure wall time?

### ISSUE THREAD

**Issue by uncttao** [OpenCilk/opencilk-project#80]
Hi, what is the equivalent of `omp_get_wtime()` in Cilk?

**Comment by neboat** [comment]
There are a couple of options for measuring wall time.
1. You can use traditional functions for measuring time, such as `clock_gettime()` from `time.h`.
2. You can alternatively use the Cilkscale API for measuring performance and compile and link your tool with a Cilkscale tool.

You can use the Cilkscale API very similarly to how you use `omp_get_wtime()`.  For example, here's how to use the Cilkscale API to measure a particular region of your program (using the C API):
```
wsp_t start, end;
start = wsp_getworkspan();
// do work to be timed
end = wsp_getworkspan();
wsp_dump(wsp_sub(end, start), "my_work");
```
**tl;dr** If you use this API, then compile and link the program with `-fcilktool=cilkscale-benchmark` to time a particular region of the program.  In this example, the timing result will appear in a CSV-formatted row labeled `my_work`.  (A different CSV-formatted row will list the timing of the whole program.)

**Long answer**: This API will measure performance in different ways, depending on how you compile and link the program.
- If you don't compile and link the program with Cilkscale --- that is, you compile and link the program normally --- then the functions `wsp_getworkspan()`, `wsp_sub()`, and `wsp_dump()` are all no-ops.
- If you compile and link the program with the Cilkscale benchmark tool --- by compiling and linking the program with `-fcilktool=cilkscale-benchmark` --- then the calls to `wsp_getworkspan()` will measure wall time, much like `omp_get_wtime()`.
- If you compile and link the program with the standard Cilkscale tool --- by compiling and linking the program with `-fcilktool=cilkscale` --- then the calls to `wsp_getworkspan()` will measure *work* and *span*, which are two performance metrics that together bound the program's parallel scalability.  You can find a tutorial description of work and span starting on slide 46 of this slide deck: https://cilk.scripts.mit.edu/pact21/opencilk-pact-2021.pdf.

If the program is compiled and linked with either the Cilkscale benchmark tool or the standard Cilkscale tool, then the `wsp_sub()` and `wsp_dump()` functions will output the measurement of the designated part of the program in a CSV-formatted labeled `my_work`.

In addition, OpenCilk comes with a Cilkscale visualizer tool that works with both the Cilkscale benchmark tool and the standard Cilkscale tool to analyze a Cilk program's scalability and benchmark its performance.  You can find more description of Cilkscale and its visualizer tool here: https://cilk.scripts.mit.edu/beta2/opencilk.spaa.2020.pdf.  In that slide deck, Cilkscale is described on slides 22-23, and a tutorial demo of using Cilkscale is given on slides 33-36.  (Note that these slides are slightly outdated, since they describe a beta version of OpenCilk.  If you have a more recent installation of OpenCilk, then you should find the Cilkscale visualizer tool in the `share/Cilkscale_vis` subdirectory of your installation.)

Let us know if you have any issues using Cilkscale and if you have any other questions.

**Comment by uncttao** [comment]
Thank you for such a detailed and helpful answer! I will take time to digest and raise more questions if any.
